ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java

```diff
@@ -0,0 +1,381 @@
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ *
+ */
+
+package org.opensearch.ml.engine.algorithms.anomalylocalization;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.NotifyOnceListener;
+import org.opensearch.action.search.MultiSearchRequest;
+import org.opensearch.action.search.MultiSearchResponse;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.Client;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.index.query.RangeQueryBuilder;
+import org.opensearch.index.query.TermQueryBuilder;
+import org.opensearch.search.aggregations.AggregationBuilder;
+import org.opensearch.search.aggregations.AggregationBuilders;
+import org.opensearch.search.aggregations.bucket.composite.CompositeAggregation;
+import org.opensearch.search.aggregations.bucket.composite.CompositeAggregationBuilder;
+import org.opensearch.search.aggregations.bucket.composite.TermsValuesSourceBuilder;
+import org.opensearch.search.aggregations.bucket.filter.Filters;
+import org.opensearch.search.aggregations.bucket.filter.FiltersAggregationBuilder;
+import org.opensearch.search.aggregations.bucket.filter.FiltersAggregator.KeyedFilter;
+import org.opensearch.search.aggregations.metrics.NumericMetricsAggregation.SingleValue;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import lombok.Data;
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+
+import static org.opensearch.action.ActionListener.wrap;
+import static org.opensearch.search.aggregations.MultiBucketConsumerService.MAX_BUCKET_SETTING;
+
+/**
+ * Implementation of AnomalyLocalizer.
+ */
+@Log4j2
+public class AnomalyLocalizerImpl implements AnomalyLocalizer {
+
+    protected static final double MIN_DIFF_PCT = 0.01;
+    protected static final int MAX_TIME_BUCKETS = 8;
+
+    private final Client client;
+    private final Settings settings;
+
+    /**
+     * Constructor.
+     *
+     * @param client   Index data.
+     * @param settings Settings information.
+     */
+    public AnomalyLocalizerImpl(
+            Client client,
+            Settings settings) {
+        this.client = client;
+        this.settings = settings;
+    }
+
+    /**
+     * Implementation of localization.
+     * <p>
+     * This localization finds the largest contributors to changes in aggregated data over time.
```

Comment:
How about add more javadoc for these key steps in different methods?

Reply:
yes, i added javadocs to the submethods. let me know if some important ones are missing or unclear.